### ISSUE TITLE
storage: Provide a WriteSeeker interface from Storage Client

### ISSUE THREAD

**Issue by noseglid** [googleapis/google-cloud-go#1243]
I am using GCS to stream audio bytes generated by [FFMpeg](https://www.ffmpeg.org/). In order to avoid storing the media on disk before uploading to GCS (it would slow down the process, and consume resources unnecessarily) I create a [`Writer`](https://godoc.org/cloud.google.com/go/storage#ObjectHandle.NewWriter). Next, I register a write function with FFMpeg, in which I simply take the data, and pass it to the `Storage Writer` which writes to GCS.

The way I register with FFmpeg is by using the [avio_alloc_context](https://www.ffmpeg.org/doxygen/trunk/avio_8h.html#a853f5149136a27ffba3207d8520172a5) which accepts a `read`, `write` and `seek` function (I'm using [cgo](https://golang.org/cmd/cgo/) to bridge between C and GO).

This all works fine, until FFMpeg wants to `seek`. Since I can't change the cursor where the bytes are written on GCS, I just pass `null` here - and FFmpeg won't do the seeking.

For audio and video media, it's very common that there are certain parts of the header which is unknown until all media is processed (such as duration of the media). This is why FFMpeg can't write the header on the first pass, and wants to seek back once it's done.

Because it can't seek, the resulting media file stored in GCS will not have those parts of the headers set, and won't work properly.

What would fully solve my problems was if there was a function on the Storage Client, `NewWriteSeeker` which returned a [WriteSeeker](https://golang.org/pkg/io/#WriteSeeker) interface. I could then leverage the `Seek` functionality of that to allow FFMpeg to do what it needs.

I know other major media libraries (such as [GStreamer](https://gstreamer.freedesktop.org/)) requires similar functionality.

I've looked at the JSON API if Cloud Storage, and this might be possible to solve with the [Resumable Upload](https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload) functionality. If not, it'd be nice if we could involve whoever can make an API satisfying these needs.

This issue was sparked from the [discussion](https://github.com/googleapis/google-cloud-go/issues/1224#issuecomment-444631926) from #1224.

**Comment by jadekler** [comment]
cc @frankyn

**Comment by odeke-em** [comment]
Sorry for the late reply @noseglid and thank you for the report! I shall ping @frankyn next week and we can brainstorm on whether this might be possible or not. It might not be directly possible with this package but I suspect Frank's advice to compose and concatenate objects could solve the issue for you but that'll require some extra work, we'll discuss anyways.

**Comment by noseglid** [comment]
Did you have a discussion yet?

**Comment by nazarslota** [comment]
Hello! I faced exactly the same problem. Are there any updates on this?

**Comment by mark-kubacki** [comment]
@noseglid Essentially the `WriteSeeker` shall overwrite bytes that have been uploaded.

With GCS that won’t work. Any bytes an _write_ or _resumable upload_ has committed to storage won’t be overwritten.

I guess you need to implement a custom buffer that’ll hold back segments you suspect will be rewritten as first step; to then utilize [composite objects](https://cloud.google.com/storage/docs/composite-objects#appends) (as did I before managing parts myself). Maybe there’s other ideas.

---
(fyi, for the other direction: A ReadSeeker is dangerous on AWS since you are actually charged for the transfer from storage to AWS’ intermediate system that actually serves your response. For small blocks, you could be charged for the entire file.)

**Comment by tritone** [comment]
Sorry for the belated response on this one. My understanding is that this request is not really compatible with GCS resumable upload APIs. For a resumable upload, the bytes must be written in order starting at offset 0. So, there is no practical way to implement this feature at the Go client layer.

I think specifically for a case like writing a header, creating two separate files and then composing them might be a viable workaround.

But yes for the general case of a Writer that can seek to arbitrary offsets, this is not possible with GCS upload APIs. You could file a feature request with your account manager but I am not aware of anything like this in the pipeline currently. We do have s3-compatible multipart upload in GCS (not available via this library yet), but even that doesn't really meet this need since chunks need to have specific memory alignments in order to be uploaded; fully arbitrary offsets are not supported.

**Comment by noseglid** [comment]
That's fair.

For anyone curious, my workaround has been to store to a local file on an ephemeral storage which supports read/write/seek. Once the ffmpeg process is complete, I upload it to GCS in full.
